lib/theme_check/language_server/variable_lookup_finder.rb

```diff
@@ -0,0 +1,263 @@
+
+        start_index = 0
+        end_index = cursor - 1
+        markup = content[start_index..end_index]
+        last_line = markup.rstrip.lines.last
+
+        is_liquid_tag = markup =~ /\A{%\s*liquid/im
+        ends_with_spaces = markup =~ / +$/
+        ends_with_potential_lookup_position = markup =~ ENDS_WITH_POTENTIAL_LOOKUP_POSITION
+
+        # Welcome to Hackcity
+        markup += "'" if markup.count("'").odd?
+        markup += '"' if markup.count('"').odd?
+        markup += "]" if markup =~ UNCLOSED_SQUARE_BRACKET
+        markup += ' %}' unless is_liquid_tag
+        markup = "{% liquid \n#{last_line}" if is_liquid_tag
+
+        # if statements
+        is_if_tag = markup =~ /\A{%\s*if/im
+        is_liquid_if = is_liquid_tag && last_line.match?(/^\s*if/)
+        return empty_lookup if (is_if_tag || is_liquid_if) && ends_with_spaces
+        markup += '{% endif %}' if is_if_tag
+        markup += "\n endif" if is_liquid_if
+
+        # unless statements
+        is_liquid_unless = is_liquid_tag && last_line.match?(/^\s*unless/)
+        is_unless_tag = markup =~ /\A{%\s*unless/im
+        return empty_lookup if (is_unless_tag || is_liquid_unless) && ends_with_spaces
+        markup += '{% endunless %}' if is_unless_tag
+        markup += "\n endunless" if is_liquid_unless
+
+        # elsif statements
+        is_elsif_tag = markup =~ /\A{%\s*elsif/im
+        is_liquid_elsif = is_liquid_tag && last_line.match?(/\A\s*elsif/)
+        return empty_lookup if (is_elsif_tag || is_liquid_elsif) && ends_with_spaces
+        markup = '{% if x %}' + markup + '{% endif %}' if is_elsif_tag
+        if is_liquid_elsif
+          markup = <<~LIQUID
+            {% liquid
+              if x
+              #{markup.lines[-1]}
+              endif
+          LIQUID
+        end
+
+        # case statements
+        is_case_tag = markup =~ /\A{%\s*case/im
+        is_liquid_case = is_liquid_tag && last_line.match?(/^\s*case/)
+        return empty_lookup if (is_case_tag || is_liquid_case) && ends_with_spaces
+        markup += "{% endcase %}" if is_case_tag
+        markup += "\n endcase" if is_liquid_case
+
+        # when
+        is_when_tag = markup =~ /\A{%\s*when/im
+        is_liquid_when = is_liquid_tag && last_line.match?(/^\s*when/)
+        return empty_lookup if (is_when_tag || is_liquid_when) && ends_with_spaces
+        markup = "{% case x %}" + markup + "{% endcase %}" if markup =~ /\A{%\s*when/im
+        markup = "{% liquid\n case x\n #{last_line}\n endcase\n" if is_liquid_when
+
+        # for statements
+        is_for_tag = markup =~ /\A{%\s*for/im
+        is_liquid_for = is_liquid_tag && last_line.match?(/^\s*(for)/)
+        return empty_lookup if (is_for_tag || is_liquid_for) && ends_with_potential_lookup_position
+        markup += "{% endfor %}" if is_for_tag
+        markup += "\n endfor" if is_liquid_for
+
+        # closing liquid tag
+        markup += "\n%}" if is_liquid_tag
```

Comment:
Wonder if it's worth it to refactor this into something a little more intelligible. I'm not sure moving the code will be any better but it's worth a try.